docs/design/2019-11-05-index-advisor.md

```diff
@@ -0,0 +1,185 @@
+
+1. **Potential indexes**. This proposal only considers single column indexes, thus all columns in a table can be potential indexes and column statistics is the corresponding index's statistics.
+
+2. **Evaluate indexes**. For single query, the optimizer will output a physical plan and its corresponding cost. We can quantify the effect of certain hypothetic indexes by compiling a query twice. In the first time, this query is compiled without adding any hypothetic indexes to *schema*. Denote the output `cost` of SQL optimizer as `original_cost`.  While in the second time, compile the same query with all potential indexes being added to *schema*. Denote indexes appeared in output execution plan as `LSI (Local Selected Index)` for this query and `reduced_cost = original_cost - virtual_cost` as `LSI`'s score where `virtual_cost` is current execution plan's cost. Simply add up each LSI's score for different queries to get the LSI's final score. Sort LSIs' final score in descending order and get a score list.
+
+3. **Evaluate index set**. Pick the top `n` indexes as the *initial recommended index set* and then launch the *Swap and Re-evaluate* algorithm. Note that executing the *Swap and Re-evaluate* algorithm is necessary as the `reduced_cost` of certain query may be **joint** effect of several indexes. Within this context, it is hard to independently estimate each index's contribution to the final `reduced_cost`. For example, assume there is an extremely slow query in input workload and the desired indexes for this query is `a` and `b`. However, the number of allowed recommended indexes for the whole workload is limited and for some reason, `a` ranks top `n` in the final score list while `b` is not. But there are chances that without `b`, `a` can no more optimize that extremely slow query. To tackle this problem, we have to *efficiently* enumerate possible optimal recommend index set for the workload. The difference from *random enumeration* is that we reasonably think the initial recommended index set composed of top `n` indexes in the score list is a *good start*. By swapping partial elements between the `recommend_set` and the `remaining_set`, we are trying *variates* of the *good start* recommend set in which we are more likely to get the real optimal recommend set as quickly as possible. The factor determining the quality of recommend index set is the overall cost of input workload after adding all indexes in recommend index set to the schema in memory. Following is snippets of pseudorandom code of `Swap and Re-evaluate` algorithm.  
```

Comment:
Can you split it into smaller logical parts. It is so long and makes me lost...

Reply:
OK.